java/core/src/main/java/com/google/protobuf/CodedInputStream.java

```diff
@@ -3851,40 +3944,16 @@ public byte readRawByte() throws IOException {
 
     @Override
     public byte[] readRawBytes(final int length) throws IOException {
-      if (length >= 0 && length <= currentRemaining()) {
-        byte[] bytes = new byte[length];
-        UnsafeUtil.copyMemory(currentByteBufferPos, bytes, 0, length);
-        currentByteBufferPos += length;
-        return bytes;
-      }
-      if (length >= 0 && length <= remaining()) {
-        byte[] bytes = new byte[length];
-        readRawBytesTo(bytes, 0, length);
-        return bytes;
-      }
-
-      if (length <= 0) {
-        if (length == 0) {
-          return EMPTY_BYTE_ARRAY;
-        } else {
-          throw InvalidProtocolBufferException.negativeSize();
-        }
+      byte[] bytes = EMPTY_BYTE_ARRAY;
+      if (length > 0 && length <= remaining()) {
+        bytes = new byte[length];
       }
-
-      throw InvalidProtocolBufferException.truncatedMessage();
+      readRawBytes(bytes, 0, length);
+      return bytes;
     }
 
-    /**
-     * Try to get raw bytes from {@code input} with the size of {@code length} and copy to {@code
-     * bytes} array. If the size is bigger than the number of remaining bytes in the input, then
-     * throw {@code truncatedMessage} exception.
-     *
-     * @param bytes
-     * @param offset
-     * @param length
-     * @throws IOException
-     */
-    private void readRawBytesTo(byte[] bytes, int offset, final int length) throws IOException {
+    @Override
+    public void readRawBytes(byte[] bytes, int offset, final int length) throws IOException {
```

Comment:
public methods must have javadoc defining what the method is doing for external users.